yt_dlp/extractor/rudovideo.py

```diff
@@ -0,0 +1,108 @@
+            'id': 'bQkt07',
+            'title': 'Tubular Bells',
+            'ext': 'mp4',
+            'thumbnail': r're:^(?:https?:)?//.*\.(png|jpg)$',
+        },
+    }, {
+        'url': 'https://rudo.video/vod/bN5AaJ',
+        'md5': '01324a329227e2591530ecb4f555c881',
+        'info_dict': {
+            'id': 'bN5AaJ',
+            'title': 'Ucrania 19.03',
+            'creator': 'La Tercera',
+            'ext': 'mp4',
+            'thumbnail': r're:^(?:https?:)?//.*\.(png|jpg)$',
+        },
+    }, {
+        'url': 'https://rudo.video/live/bbtv',
+        'info_dict': {
+            'id': 'bbtv',
+            'ext': 'mp4',
+            'creator': 'BioBioTV',
+            'live_status': 'is_live',
+            'title': r're:^LIVE BBTV\s\d{4}-\d{2}-\d{2}\s\d{2}:\d{2}$',
+            'thumbnail': r're:^(?:https?:)?//.*\.(png|jpg)$',
+        },
+    }, {
+        'url': 'https://rudo.video/live/c13',
+        'info_dict': {
+            'id': 'c13',
+            'title': 'CANAL13',
+            'ext': 'mp4',
+        },
+        'skip': 'Geo-restricted to Chile',
+    }, {
+        'url': 'https://rudo.video/live/t13-13cl',
+        'info_dict': {
+            'id': 't13-13cl',
+            'title': 'T13',
+            'ext': 'mp4',
+        },
+        'skip': 'Geo-restricted to Chile',
+    }]
+
+    def _real_extract(self, url):
+        video_id, type_ = self._match_valid_url(url).group('id', 'type')
+        is_live = type_ == 'live'
+
+        webpage = self._download_webpage(url, video_id)
+        if 'Streaming is not available in your area' in webpage:
+            self.raise_geo_restricted()
+
+        m3u8_url = (
+            self._search_regex(
+                r'var\s+streamURL\s*=\s*[\'"]([^?\'"]+)', webpage, 'stream url', default=None)
+            # Source URL must be used only if streamURL is unavailable
+            or self._search_regex(
+                r'<source[^>]+src=[\'"]([^\'"]+)', webpage, 'source url', default=None))
+        if not m3u8_url:
+            youtube_url = self._search_regex(r'file:\s*[\'"]((?:https?:)//(?:www\.)?youtube\.com[^\'"]+)',
+                                             webpage, 'youtube url', default=None)
+            if youtube_url:
+                return self.url_result(youtube_url, 'Youtube')
+            raise ExtractorError('Unable to extract stream url')
+
+        token_array = self._search_json(
+            r'<script>var\s+_\$_[a-zA-Z0-9]+\s*=', webpage, 'access token array', video_id,
+            contains_pattern=r'\[(?s:.+)\]', default=None, transform_source=js_to_json)
+        if token_array:
+            if len(token_array) != 9:
+                raise ExtractorError('Invalid access token array')
+            access_token = self._download_json(token_array[0], video_id, note='Downloading access token')
+            m3u8_url = update_url_query(m3u8_url, {
+                'auth-token': traverse_obj(access_token, ('data', 'authToken'))
+            })
```

Comment:
should this be fatal?
```suggestion
            m3u8_url = update_url_query(m3u8_url, {
                'auth-token': access_token['data']['authToken'],
            })
```

Reply:
Think yes, it should be fatal.
It would mean that there is a token array, and called the token API, but the response didn't have the required token.
What's the correct way of handling that? Should I change it as you suggested? Or should I validate that `access_token` contains `data` and `authToken`, and raise `ExtractorError` if it doesn't have them?

Also, now I saw that there is a `if len(token_array) != 9:` ... that was because that array has several members, and there are 2 members with the `auth-token` and `authToken` strings. My first PR instead of using `authToken` and `auth-token`, it used the `token_array` elements. 

I think it looks better to remove that check , or to check that it has at least 1 element (assuming it's the token's API endpoint).

What do you think?

Reply:
I would do this instead
```py
        if token_array:
            token_url = traverse_obj(token_array, (..., {url_or_none}), get_all=False)
            if not token_url:
                raise ExtractorError('Invalid access token array')
            access_token = self._download_json(
                token_url, video_id, note='Downloading access token')['data']['authToken']
            m3u8_url = update_url_query(m3u8_url, {'auth-token': access_token})
```

Reply:
ah alright. LGTM then